pkg/3scale/amp/operator/system_reconciler.go

```diff
@@ -182,6 +143,134 @@ func (r *SystemReconciler) Reconcile() (reconcile.Result, error) {
 		return reconcile.Result{}, err
 	}
 
+	// Used to synchronize rollout of system Deployments
+	systemComponentNotReady := false
+
+	// If the system-app Deployment generation has changed, delete the PreHook/PostHook Jobs so they can be recreated
+	generationChanged, err := helper.HasAppGenerationChanged(component.SystemAppPreHookJobName, component.SystemAppDeploymentName, r.apiManager.GetNamespace(), r.Client())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if generationChanged {
+		err = helper.DeleteJob(component.SystemAppPreHookJobName, r.apiManager.GetNamespace(), r.Client())
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+	generationChanged, err = helper.HasAppGenerationChanged(component.SystemAppPostHookJobName, component.SystemAppDeploymentName, r.apiManager.GetNamespace(), r.Client())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if generationChanged {
+		err = helper.DeleteJob(component.SystemAppPostHookJobName, r.apiManager.GetNamespace(), r.Client())
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// Used to synchronize the system-app Deployment with the PreHook/PostHook Jobs
+	currentAppDeploymentGeneration, err := getSystemAppDeploymentGeneration(r.apiManager.GetNamespace(), r.Client())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// SystemApp PreHook Job
+	preHookJob := system.AppPreHookJob(ampImages.Options.SystemImage, currentAppDeploymentGeneration)
+	err = r.ReconcileJob(preHookJob, reconcilers.CreateOnlyMutator)
```

Comment:
So during 2.14 > 2.15, we will run the prehook twice right? 
First, on initial when we hit this line of code (since no jobs are present)
Second, after updating the deployment via deployment mutators?

Reply:
/No the PreHook will only run once (on initial run since Job doesn't exist). Subsequent reconciles won't recreate/run the Job since it already exists. The exception is if the `system-app` Deployment's generation changes which will cause the PreHook Job to be deleted and then recreated.

Reply:
ah yes, I see, so we are returning "1" generation in AppPreHookJob on initial reconcile in upgrade scenario but then create the deployment - with generation 1. So the second job doesn't trigger because the generation is the same. Thanks